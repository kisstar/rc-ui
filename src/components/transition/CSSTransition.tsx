import React, { useRef, useContext, useMemo, useCallback } from 'react';
import { CSSTransition as ReactCSSTransition } from 'react-transition-group';
import { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import { ConfigContext } from '../config-provider';

type AnimationName = string;

type AnimationTuple = [AnimationName, AnimationName];

interface OwnProps {
  animation?: AnimationName | AnimationTuple;
  prefixCls?: string;
}

export type TransitionProps<T extends HTMLElement = HTMLElement> = OwnProps & CSSTransitionProps<T>;

const CSSTransition: React.FC<TransitionProps> = props => {
  const {
    children,
    classNames,
    animation,
    nodeRef,
    timeout = 0,
    prefixCls: customizePrefixCls,
    ...restProps
  } = props;
  const defaultNodeRef = useRef(null);
  const { getPrefixCls } = useContext(ConfigContext);

  const transitionClasses = useMemo(() => {
    if (!animation) return {};

    let retAnimation: AnimationTuple;
    const prefixCls = getPrefixCls('animate', customizePrefixCls);

    if (!Array.isArray(animation)) {
      retAnimation = [animation, animation];
    } else {
      retAnimation = animation as AnimationTuple;
    }

    return {
      enter: `${prefixCls}-animated`,
      enterActive: `${prefixCls}-${retAnimation[0]}`,
      exit: `${prefixCls}-animated`,
      exitActive: `${prefixCls}-${retAnimation[1]}`,
    };
  }, [animation, customizePrefixCls, getPrefixCls]);

  const getChildProps = useCallback(() => {
    const childProps = {};
    if (!nodeRef) {
      Object.assign(childProps, {
        ref: defaultNodeRef,
      });
    }
    if (animation) {
      const duration = `${timeout}ms`;

      Object.assign(childProps, {
        style: {
          animationDuration: duration,
          WebkitAnimationDuration: duration,
        },
      });
    }
    return childProps;
  }, [timeout, animation, nodeRef]);

  return (
    <ReactCSSTransition
      timeout={timeout}
      nodeRef={nodeRef || defaultNodeRef}
      classNames={animation ? transitionClasses : classNames}
      {...restProps}
    >
      {React.cloneElement(
        children as React.FunctionComponentElement<TransitionProps>,
        getChildProps(),
      )}
    </ReactCSSTransition>
  );
};

CSSTransition.defaultProps = {
  unmountOnExit: true,
};

export default CSSTransition;
