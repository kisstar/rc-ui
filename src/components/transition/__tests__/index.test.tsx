import React from 'react';
import { render, RenderResult } from '@testing-library/react';
// import { CSSTransition } from 'react-transition-group';
import mountTest from '../../../tests/mountTest';
import { getStyleStr } from '../../../tests/utils';
import { CSSTransition } from '..';
import { TransitionProps } from '../CSSTransition';

const prefixCls = 'ks-animate';

describe('CSSTransition', () => {
  it('renders correctly', () => {
    expect(
      <CSSTransition timeout={1000}>
        <div>I will receive fade-* classes</div>
      </CSSTransition>,
    ).toMatchSnapshot();
  });

  mountTest(() => (
    <CSSTransition timeout={1000} animation="flash">
      <div className="box" />
    </CSSTransition>
  ));

  it('should apply classes at each transition state', async done => {
    let count = 0; // make sure that each hook is called
    let resolve: () => void;
    const p = new Promise(res => {
      resolve = res;
    });
    const nodeRef = React.createRef<HTMLDivElement>();
    const initProps = {
      in: false,
      nodeRef,
      classNames: 'test',
      timeout: 10,
    };
    const newProps = {
      ...initProps,
      in: true,

      onEnter() {
        count += 1;
        expect((nodeRef.current as HTMLDivElement).className).toEqual('test-enter');
      },

      onEntering() {
        count += 1;
        expect((nodeRef.current as HTMLDivElement).className).toEqual(
          'test-enter test-enter-active',
        );
      },

      onEntered() {
        count += 1;
        expect((nodeRef.current as HTMLDivElement).className).toEqual('test-enter-done');
        resolve();
      },
    };
    const retProps = {
      ...initProps,

      onExit() {
        count += 1;
        expect((nodeRef.current as HTMLDivElement).className).toEqual('test-exit');
      },

      onExiting() {
        count += 1;
        expect((nodeRef.current as HTMLDivElement).className).toEqual('test-exit test-exit-active');
      },

      onExited() {
        expect((nodeRef.current as HTMLDivElement).className).toEqual('test-exit-done');
        expect(count).toEqual(5);
        done();
      },
    };
    const TestComponent = (props: TransitionProps) => (
      <CSSTransition {...props}>
        <div ref={nodeRef} />
      </CSSTransition>
    );
    const wrapper = render(<TestComponent {...initProps} />);

    wrapper.rerender(<TestComponent {...newProps} />);
    await p;
    wrapper.rerender(<TestComponent {...retProps} />);
  });

  it('should automatically create noderef and pass it to origin CSSTransition', done => {
    const initProps = {
      in: false,
      classNames: 'test',
      timeout: 10,
    };
    const newProps = {
      ...initProps,
      in: true,

      onEntered() {
        done();
      },
    };
    const TestComponent = (props: TransitionProps) => (
      <React.StrictMode>
        <CSSTransition {...props}>
          <div>Text</div>
        </CSSTransition>
      </React.StrictMode>
    );
    const wrapper = render(<TestComponent {...initProps} />);

    // 由于 React 并没有将在严格模式下使用 findDOMNode 方法视为错误（通过 console.error 提示），所以这个的期望永远可以通过
    // 如果这个 case 失败，你将在控制台看到相应的提示信息
    expect(() => {
      wrapper.rerender(<TestComponent {...newProps} />);
    }).not.toThrow();
  });

  it('should support the animation provided by animate.css', async done => {
    // prefixCls
    let count = 0; // make sure that each hook is called
    let resolve: () => void;
    const p = new Promise(res => {
      resolve = res;
    });
    let wrapper: RenderResult;
    let element: HTMLElement;
    const initProps = {
      in: false,
      timeout: 10,
    };
    const testStyle = {
      animationDuration: `${initProps.timeout}ms`,
    };
    const newProps = {
      ...initProps,
      in: true,

      onEnter() {
        element = wrapper.getByText('Text');
        count += 1;
        expect(element.className).toEqual(`${prefixCls}-animated`);
        expect(element).toHaveStyle(getStyleStr(testStyle));
      },

      onEntering() {
        count += 1;
        expect(element.className).toEqual(`${prefixCls}-animated ${prefixCls}-testIn`);
        expect(element).toHaveStyle(getStyleStr(testStyle));
      },

      onEntered() {
        count += 1;
        expect(element.className).toEqual('');
        expect(element).toHaveStyle(getStyleStr(testStyle));
        resolve();
      },
    };
    const retProps = {
      ...initProps,
      in: false,

      onExit() {
        count += 1;
        expect(element.className).toEqual(`${prefixCls}-animated`);
        expect(element).toHaveStyle(getStyleStr(testStyle));
      },

      onExiting() {
        count += 1;
        expect(element.className).toEqual(`${prefixCls}-animated ${prefixCls}-testOut`);
        expect(element).toHaveStyle(getStyleStr(testStyle));
      },

      onExited() {
        expect(element.className).toEqual('');
        expect(element).toHaveStyle(getStyleStr(testStyle));
        expect(count).toEqual(5);
        done();
      },
    };
    const TestComponent = (props: TransitionProps) => (
      <CSSTransition {...props} animation={['testIn', 'testOut']}>
        <div>Text</div>
      </CSSTransition>
    );

    wrapper = render(<TestComponent {...initProps} />);
    wrapper.rerender(<TestComponent {...newProps} />);
    await p;
    wrapper.rerender(<TestComponent {...retProps} />);
  });
});
