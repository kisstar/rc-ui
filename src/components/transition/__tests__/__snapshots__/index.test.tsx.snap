// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSSTransition renders correctly 1`] = `
<CSSTransition
  timeout={1000}
  unmountOnExit={true}
>
  <div>
    I will receive fade-* classes
  </div>
</CSSTransition>
`;
